Add CSI block volume directory cleanup

jsafrane opened the pull request:
CSI volume plugin creates number of files/directories when processing block volumes. These files must be cleaned when the plugin is done with the volume, i.e. at the end on `TearDownDevice` or when `SetupDevice` fails.

I added unit tests covering both `SetupDevice` failure and full lifetime of a CSI block volume from `SetupDevice` trough `MapPodDevice`, `UnmapPodDevice` and `TearDownDevice` to check that there are no orphan files.

Some changes to fake CSI client were necessary to create / delete publish path by the fake driver.

/kind bug

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #87977 #79780

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed cleaning of CSI raw block volumes.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

@msau42 @gnufied

jsafrane commented:
cc @mkimuram for review how the directories should work

davidz627 commented:
/assign @misterikkit

jsafrane commented:
/retest

gnufied commented on pkg/volume/csi/csi_block.go:
```diff
@@ -299,6 +301,13 @@ func (m *csiBlockMapper) SetUpDevice() error {
 	// Call NodeStageVolume
 	_, err = m.stageVolumeForBlock(ctx, csiClient, accessMode, csiSource, attachment)
 	if err != nil {
+		if volumetypes.IsOperationFinishedError(err) {
```
Do we need to introduce uncertain device tracking for block volume calls? We can do this in a follow up but we should track that work somewhere?

jsafrane replied:
Yes, we should. Filled https://github.com/kubernetes/kubernetes/pull/88016

gnufied replied:
Wait is that correct? That is another PR..

jsafrane replied:
sorry, I fixed my previous comment with the right link

mkimuram commented on pkg/volume/csi/csi_block.go:
```diff
@@ -435,6 +444,57 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+func (m *csiBlockMapper) cleanupOrphanDeviceFiles() error {
+	// Remove artifacts of NodePublish.
+	// publishPath: xxx/plugins/kubernetes.io/csi/volumeDevices/publish/<volume name>/<pod UUID>
+	// publishPath was removed by the driver. We need to remove the <volume name>/ dir.
```
Correct. And this seems the right place to delete parent path of publish path for a volume, because we are not sure that we can delete it until the volume is no longer published for any pods. (I first thought that it should be handled in `MapPodDevice`/`UnmapPodDevice`, but it's not true.)

mkimuram commented on pkg/volume/csi/csi_block.go:
```diff
@@ -435,6 +444,57 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+	publishDir := filepath.Dir(publishPath)
+	if m.podUID == "" {
+		// Pod UID is not known during device teardown ("NodeUnstage").
+		// getPublishPath() squashed "<volume name>/<pod UUID>" into "<volume name>/".
```
Right. This is needed for this function to be called on error in `SetUpDevice`.

mkimuram commented on pkg/volume/csi/csi_block.go:
```diff
@@ -435,6 +444,57 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
 		}
 	}
+	if err = m.cleanupOrphanDeviceFiles(); err != nil {
+		return err
```
Should we keep trying to delete directories, here? Or can we just log and return without error as it is done in `SetUpDevice`?

mkimuram commented on pkg/volume/csi/csi_block.go:
```diff
@@ -435,6 +444,57 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+	}
+
+	// Remove everything under xxx/plugins/kubernetes.io/csi/volumeDevices/<volume name>.
+	// At this point it contains only "data/vol_data.json" and empty "dev/".
```
This is right. I seemed to forget to write down the existence of "data/vol_data.json" under xxx/plugins/kubernetes.io/csi/volumeDevices/<volume name> in L56 comments. 

Should we add a comment on the file, there?
Then, I think that everyone will be able to check that all paths are cleaned up properly, just by comparing the comments and this cleanup function.

mkimuram reviewed (commented):
@jsafrane 

Thank you for working on this. From the viewpoint of directories, it looks good.

misterikkit commented on pkg/volume/csi/fake/fake_client.go:
```diff
@@ -172,14 +175,29 @@ func (f *NodeClient) NodePublishVolume(ctx context.Context, req *csipb.NodePubli
 		Path:            req.GetTargetPath(),
 		DeviceMountPath: req.GetStagingTargetPath(),
 		VolumeContext:   req.GetVolumeContext(),
 		FSType:          req.GetVolumeCapability().GetMount().GetFsType(),
```
You conditionally added `MountFlags`. Should `FSType` also be made conditional? (not worried about the nil check, just the logic.)

Or is the difference between empty string and nil `[]string` significant in the way we do `CSIVolume` objects?

jsafrane replied:
Fixed (moved FsType under `if GetMount() != nil`)

jsafrane commented on pkg/volume/csi/csi_block.go:
```diff
@@ -435,6 +444,57 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+		// getPublishPath() squashed "<volume name>/<pod UUID>" into "<volume name>/".
+		publishDir = publishPath
+	}
+	if err := os.Remove(publishDir); err != nil && !os.IsNotExist(err) {
```
`Remove` is on the safe side. With our usage of bind mounts I don't want to accidentally remove stuff that was bind-mounted to a wrong place.

jsafrane commented on pkg/volume/csi/fake/fake_client.go:
```diff
@@ -172,14 +175,29 @@ func (f *NodeClient) NodePublishVolume(ctx context.Context, req *csipb.NodePubli
+	// "Creation of target_path is the responsibility of the SP."
+	// Our plugin depends on it.
+	if req.VolumeCapability.GetBlock() != nil {
+		if err := ioutil.WriteFile(req.TargetPath, []byte{}, 0644); err != nil {
```
Yes.

davidz627 commented:
/uncc

jsafrane commented on pkg/volume/csi/csi_block.go:
```diff
@@ -435,6 +444,57 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+	// publishPath was removed by the driver. We need to remove the <volume name>/ dir.
+	publishPath := m.getPublishPath()
+	publishDir := filepath.Dir(publishPath)
+	if m.podUID == "" {
```
I agree that the code assumes too much about paths created at various places of the plugin. Do you have a better suggestion?

msau42 replied:
would it help to have helper methods to extract various parts of the path?

jsafrane replied:
Indeed, it looks better with `getPublishDir` and `getVolumePluginDir`.

jsafrane commented on pkg/volume/csi/csi_block.go:
```diff
@@ -435,6 +444,57 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+		publishDir = publishPath
+	}
+	if err := os.Remove(publishDir); err != nil && !os.IsNotExist(err) {
+		return errors.New(log("failed to publish directory [%s]: %v", publishDir, err))
```
fixed

jsafrane commented on pkg/volume/csi/csi_block.go:
```diff
@@ -435,6 +444,57 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+		return errors.New(log("failed to delete volume staging path [%s]: %v", stagingPath, err))
+	}
+
+	// Remove everything under xxx/plugins/kubernetes.io/csi/volumeDevices/<volume name>.
```
RemoveAll would remove also stuff that should be removed by GenerateUnmapVolumeFunc and we would not see this bug.

jsafrane replied:
Ok, I added `RemoveAll()` there, this directory contains only "global" bind-mount and metadata json. It is not used in stage/publish/unpublish/unstage loop, where blind RemoveAll could hide hard to track unpublish / unstage errors.

misterikkit replied:
Thanks, I didn't realize the subtle implications with reference to bind mount and GenerateUnmapVolumeFunc.

jsafrane commented on pkg/volume/csi/csi_block_test.go:
```diff
@@ -282,6 +283,54 @@ func TestBlockMapperSetupDevice(t *testing.T) {
+	fClient := csiMapper.csiClient.(*fakeCsiDriverClient)
+	fClient.nodeClient.SetNextError(errors.New("mock final error"))
+
+	attachID := getAttachmentName(csiMapper.volumeID, string(csiMapper.driverName), string(nodeName))
```
Fixed

jsafrane commented on pkg/volume/csi/csi_block_test.go:
```diff
@@ -282,6 +283,54 @@ func TestBlockMapperSetupDevice(t *testing.T) {
+	attachID := getAttachmentName(csiMapper.volumeID, string(csiMapper.driverName), string(nodeName))
+	attachment := makeTestAttachment(attachID, nodeName, pvName)
+	attachment.Status.Attached = true
+	_, err = csiMapper.k8s.StorageV1().VolumeAttachments().Create(context.TODO(), attachment, metav1.CreateOptions{})
```
added Background() to both places in csi_block_test.go. There is IMO no need to do anything fancier, unit tests should already have a timeout.

misterikkit replied:
I agree.

jsafrane commented on pkg/volume/csi/csi_block_test.go:
```diff
@@ -430,3 +479,124 @@ func TestBlockMapperTearDownDevice(t *testing.T) {
+	}
+	t.Log("created attachement ", attachID)
+
+	// SetupDevice
```
I'm removing the comment (and similar ones).

jsafrane commented on pkg/volume/csi/csi_block_test.go:
```diff
@@ -430,3 +479,124 @@ func TestBlockMapperTearDownDevice(t *testing.T) {
+		t.Errorf("csi server expected device path %s, got %s", stagingPath, svol.Path)
+	}
+
+	// MapPodDevice
```
These steps are noted on the top of csi_block.go. Some steps are done in operation_executor (i.e. out of scope of this unit test) or just Get* functions that are not so interesting in this test. Goal is to check that volume setup / teardown does not leave orphan files.

I added a comment on top of the test func.

jsafrane commented on pkg/volume/csi/csi_client_test.go:
```diff
@@ -385,6 +401,13 @@ func TestClientNodeGetInfo(t *testing.T) {
 }
 
 func TestClientNodePublishVolume(t *testing.T) {
+	tmpDir, err := utiltesting.MkTmpdir("csi-test")
```
The tests run in series right now, so I would leave it up to the guy who makes them parallel :-). Slightly more complicate refactoring would be needed then, as the target path is part of the test struct before any `t.Run`.

jsafrane committed:
Fix unit tests

jsafrane committed:
Add CSI block volume directory cleanup

CSI volume plugin creates number of files/directories when processing block
volumes. These files must be cleaned when the plugin is done with the
volume, i.e. at the end on TearDownDevice().

jsafrane commented:
@misterikkit, I responded to your comments and fixed some :-), can you please take another look?

jsafrane committed:
Add getPublishDir and getVolumePluginDir

So we don't need to compute these backwards from getPublishPath and
getVolumeDevicePluginDir.

jsafrane commented:
With one RemoveAll and helper `get` functions it looks much nicer now. @misterikkit PTAL.

misterikkit reviewed (commented):
For all the parts that I currently understand,
/lgtm
